### PR TITLE
PERF/BUG: ensure we store contiguous arrays in DataFrame(ndarray) for ArrayManager

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
https://github.com/pandas-dev/pandas/pull/42689 removed an "unwanted" copy, but I actually added this on purpose. This is to ensure we store contiguous 1D arrays by default, which is important for performance reasons.

### CODE REVIEW

File: pandas/tests/frame/methods/test_values.py [hunk 1]
```diff
@@ -226,7 +226,10 @@ def test_values_lcd(self, mixed_float_frame, mixed_int_frame):
 
 
 class TestPrivateValues:
-    def test_private_values_dt64tz(self, request):
```
Comment (jbrockmendel) [inline pandas/tests/frame/methods/test_values.py:229]: on the off-chance this PR isn't merged, should remove the request arg here

File: pandas/tests/frame/test_constructors.py [hunk 2]
```diff
@@ -256,12 +256,17 @@ def test_constructor_dtype_nocast_view_dataframe(self):
         should_be_view[0][0] = 99
         assert df.values[0, 0] == 99
 
-    @td.skip_array_manager_invalid_test  # TODO(ArrayManager) keep view on 2D array?
```
Comment (jbrockmendel) [inline pandas/tests/frame/test_constructors.py:259]: > keep view on 2D array?

is this no longer a goal?

File: pandas/core/internals/construction.py [hunk 3]
```diff
@@ -349,14 +349,15 @@ def ndarray_to_mgr(
...
+                    maybe_infer_to_datetimelike(values[:, i].copy())
                 )
                 for i in range(values.shape[1])
             ]
         else:
             if is_datetime_or_timedelta_dtype(values.dtype):
                 values = ensure_wrapped_if_datetimelike(values)
-            arrays = [values[:, i] for i in range(values.shape[1])]
+            # copy the array to ensure contiguous memory
+            arrays = [values[:, i].copy() for i in range(values.shape[1])]
```
Comment (jbrockmendel) [inline pandas/core/internals/construction.py:360]: could do the copy+comment once on the next line instead of both here and L352

**Comment by jbrockmendel** [comment]
this means we're not respecting the copy=False passed to DataFrame constructor?

**Comment by jorisvandenbossche** [comment]
> this means we're not respecting the copy=False passed to DataFrame constructor?

Yes, and that was on purpose. Note that the intent was to have this as the _default_ for now, and we can see later if we can provide an (optional) optimization for people that care about preserving a view on the 2D array.

Having this as the default for now also makes it easier to compare benchmarks, otherwise we would need to update all ASVs to add a `copy()` if a dataframe is created from an ndarray.

Also in any case, `df.values` will likely never be a view on the original ndarray `arr` in the standard case, even if `DataFrame(arr)` preserves a view on `arr` (because otherwise we would need to detect that each array is a subsequent slice of a single parent array, which I don't think we want to start doing). So in that sense, I think the `# TODO(ArrayManager) keep view on 2D array?` could be removed anyway now I updated the test to use multiple columns (and not be about the 1-column special case)

File: pandas/core/internals/construction.py (hunk 3 above)
Comment (jorisvandenbossche) [reply pandas/core/internals/construction.py:360]: I think that would require an additional list comprehension over the arrays?

**Comment by jorisvandenbossche** [comment]
We talked before (on the mailing list discussion about BlockManager/ArrayManager) about a potential way to keep `DataFrame(array).values` a fast roundtrip for the ndarray (eg by delaying creating the manager in this case), which would be useful for eg scikit-learn. But I see that as a future potential optimization to look into, if this turns out to be a bottleneck / import use case.

BTW, I could already add some logic to honor an explicit `copy=False` (with the default `copy=None` preserve the "ensure contiguous 1D array" behaviour)

**Comment by jbrockmendel** [comment]
>> this means we're not respecting the copy=False passed to DataFrame constructor?

> Yes, and that was on purpose. Note that the intent was to have this as the default for now, and we can see later if we can provide an (optional) optimization for people that care about preserving a view on the 2D array.

If that's the case, then it can/should be treated the same way as we currently treat dicts passed to the DataFrame constructor (where the default is copy=None).  i.e. it should be made explicit and documented.

**Comment by jorisvandenbossche** [comment]
> If that's the case, then it can/should be treated the same way as we currently treat dicts passed to the DataFrame constructor (where the default is copy=None). i.e. it should be made explicit and documented.

@jbrockmendel updated the PR to handle this through `copy=None` which is translated into `copy=True` if using array manager and input is ndarray, so using the same way how this keyword is updated for dicts.

I didn't add this to the public _docstring_ of DataFrame because currently ArrayManager is not yet documented in general, but it's now documented for devs in the comments of the the `copy=None` handling.

File: pandas/core/internals/construction.py [hunk 4]
```diff
@@ -349,14 +350,19 @@ def ndarray_to_mgr(
         if dtype is None and is_object_dtype(values.dtype):
             arrays = [
                 ensure_wrapped_if_datetimelike(
-                    maybe_infer_to_datetimelike(values[:, i])
+                    maybe_infer_to_datetimelike(values[:, i].copy())
                 )
                 for i in range(values.shape[1])
             ]
+            if copy:
```
Comment (jreback) [inline pandas/core/internals/construction.py:357]: i would move the copy *after* the main if/else here for clarify

**Review by jreback — CHANGES_REQUESTED** [review]
can you verify that the existing BM code paths are unchanged by this in perf (e.g. run some asv's for construction)

File: pandas/core/internals/construction.py (hunk 4 above)
Comment (jorisvandenbossche) [reply pandas/core/internals/construction.py]: Yes, good idea, updated

**Comment by jorisvandenbossche** [comment]
> can you verify that the existing BM code paths are unchanged by this in perf 

So in the constructor, the only thing that changes in the BM path is one additional `elif manager == "array"` check, which should not have significant performance impact.

File: pandas/core/frame.py [hunk 5]
```diff
@@ -618,6 +611,21 @@ def __init__(
 
         manager = get_option("mode.data_manager")
 
+        if copy is None:
+            if isinstance(data, dict) or data is None:
```
Comment (jbrockmendel) [inline pandas/core/frame.py:615]: at this point we shouldn't have `data is None`

File: pandas/core/internals/construction.py [hunk 6]
```diff
@@ -328,8 +328,9 @@ def ndarray_to_mgr(
         # GH#40110 see similar check inside sanitize_array
         rcf = not (is_integer_dtype(dtype) and values.dtype.kind == "f")
 
+        copy_on_sanitize = False if typ == "array" else copy
```
Comment (jbrockmendel) [inline pandas/core/internals/construction.py:331]: could define this just once earlier on?

**Comment by jbrockmendel** [comment]
couple small comments, otherwise looks good

**Comment by jorisvandenbossche** [comment]
The one failure is an unexpected ResourceWarning

File: pandas/tests/frame/methods/test_values.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/frame/methods/test_values.py:229]: using_array_manager isnt used?

Comment (jorisvandenbossche) [reply pandas/tests/frame/methods/test_values.py:229]: Yep, and request also not anymore. Updated

**Pull request merged**